### PR TITLE
Void TryCatch

### PR DESCRIPTION
**Pull request by bchong95**
# Void TryCatch

## Description

`TryCatch<TResult>` always assumes that the monad returns a result. This is not true for functions that return void. The common solution for this is to create another struct called `TryCatch` that doesn't have a generic and only works on void functions. This has the unfortunate implication of doubling all the code. `Func<T>` and `Action` are good examples of this "antipattern". One solution is to use `TryCatch<object>` and just return `null` as a placeholder, but that is syntactically misleading. The solution is to introduce a `Void` type so that you can do `TryCatch<Void>` which is more syntactically meaningful. You can take it a step further and create a `TryCatch` which is just syntactic sugar on top of `TryCatch<Void>` (which is just syntatic sugar on top of `Either<Exception, Void>`) and that way the implementation stays in sync.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/Core/Monads/TryCatch.cs [hunk 1]
```diff
@@ -183,4 +182,100 @@ public static TryCatch<TResult> FromException(Exception exception)
...
+        {
+            return new TryCatch(TryCatch<Void>.FromException(exception));
+        }
+
+        /// <summary>
+        /// Represents a void return type.
+        /// </summary>
+        private readonly struct Void
+        {
+            public static readonly Void Singleton = new Void();
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Monads/TryCatch.cs]: >public static readonly Void Singleton = new Void(); [](start = 12, length = 51)

no need for this

File: Microsoft.Azure.Cosmos/src/Query/Core/Monads/TryCatch.cs [hunk 2]
```diff
@@ -183,4 +182,99 @@ public static TryCatch<TResult> FromException(Exception exception)
             return new TryCatch<TResult>(exception);
         }
     }
+
+    internal readonly struct TryCatch
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/Monads/TryCatch.cs]: Should these be in separate files? TryCatchOfT.cs and TryCatch.cs

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs [hunk 3]
```diff
@@ -315,15 +315,15 @@ public override void Stop()
         /// <param name="tryFilterAsync">The callback used to filter each partition.</param>
         /// <param name="cancellationToken">The cancellation token.</param>
         /// <returns>A task to await on.</returns>
-        protected async Task<TryCatch<bool>> TryInitializeAsync(
+        protected async Task<TryCatch> TryInitializeAsync(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]: Does this fix a bug a user can hit?

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]: No it just makes the code a little more readable

**Pull request merged**